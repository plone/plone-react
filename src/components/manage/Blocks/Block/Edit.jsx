/**
 * Edit block.
 * @module components/manage/Blocks/Block/Edit
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { DragSource, DropTarget } from 'react-dnd';
import { findDOMNode } from 'react-dom';
import { defineMessages, injectIntl } from 'react-intl';
import { blocks } from '~/config';
import { Button } from 'semantic-ui-react';
import includes from 'lodash/includes';
import cx from 'classnames';

import withObjectBrowser from '../../Sidebar/ObjectBrowser';
import Icon from '../../../../components/theme/Icon/Icon';
import dragSVG from '../../../../icons/drag.svg';
import trashSVG from '../../../../icons/delete.svg';

const messages = defineMessages({
  unknownBlock: {
    id: 'Unknown Block',
    defaultMessage: 'Unknown Block {block}',
  },
});

const itemSource = {
  beginDrag(props) {
    return {
      id: props.id,
      index: props.index,
    };
  },
};

const ItemTypes = {
  ITEM: 'block',
};

const itemTarget = {
  hover(props, monitor, component) {
    const dragIndex = monitor.getItem().index;
    const hoverIndex = props.index;

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = findDOMNode(component).getBoundingClientRect();

    // Get vertical middle
    const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    // Get pixels to the top
    const hoverClientY = clientOffset.y - hoverBoundingRect.top;

    // Only perform the move when the mouse has crossed half of the items height
    // When dragging downwards, only move when the cursor is below 50%
    // When dragging upwards, only move when the cursor is above 50%

    // Dragging downwards
    if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
      return;
    }

    // Dragging upwards
    if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
      return;
    }

    // Time to actually perform the action
    props.onMoveBlock(dragIndex, hoverIndex);

    // Note: we're mutating the monitor item here!
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    monitor.getItem().index = hoverIndex;
  },
};

/**
 * Edit block class.
 * @class Edit
 * @extends Component
 */
class Edit extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    type: PropTypes.string.isRequired,
    data: PropTypes.objectOf(PropTypes.any).isRequired,
    selected: PropTypes.bool.isRequired,
    connectDragSource: PropTypes.func.isRequired,
    connectDragPreview: PropTypes.func.isRequired,
    connectDropTarget: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired,
    id: PropTypes.string.isRequired,
    onMoveBlock: PropTypes.func.isRequired,
    onDeleteBlock: PropTypes.func.isRequired,
  };

  componentDidMount() {
    const { type } = this.props;
    const blockHasOwnFocusManagement =
      blocks.blocksConfig?.[type]?.['blockHasOwnFocusManagement'] || null;
    if (
      !blockHasOwnFocusManagement &&
      this.props.selected &&
      this.blockNode.current
    ) {
      this.blockNode.current.focus();
    }
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const { selected, type } = this.props;
    const blockHasOwnFocusManagement =
      blocks.blocksConfig?.[type]?.['blockHasOwnFocusManagement'] || null;
    if (
      !blockHasOwnFocusManagement &&
      selected !== nextProps.selected &&
      this.blockNode.current
    ) {
      this.blockNode.current.focus();
    }
  }

  blockNode = React.createRef();

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    const {
      id,
      type,
      selected,
      connectDragSource,
      connectDropTarget,
      connectDragPreview,
    } = this.props;

    const Block = blocks.blocksConfig?.[type]?.['edit'] || null;
    const blockHasOwnFocusManagement =
      blocks.blocksConfig?.[type]?.['blockHasOwnFocusManagement'] || null;

    const hideHandler =
      this.props.data['@type'] === 'text' &&
      (!this.props.data.text ||
        (this.props.data.text &&
          this.props.data.text.blocks &&
          this.props.data.text.blocks.length === 1 &&
          this.props.data.text.blocks[0].text === ''));

    return connectDropTarget(
      connectDragPreview(
        <div className={`ui drag block inner ${type}`}>
          {selected &&
            connectDragSource(
              <div
                className={
                  hideHandler
                    ? 'drag handle wrapper hidden'
                    : 'drag handle wrapper'
                }
              >
                <Icon className="drag handle" name={dragSVG} size="18px" />
              </div>,
            )}
          {Block !== null ? (
            <div
              role="presentation"
              onClick={() => this.props.onSelectBlock(this.props.id)}
              onKeyDown={
                !blockHasOwnFocusManagement
                  ? e =>
                      this.props.handleKeyDown(
                        e,
                        this.props.index,
                        this.props.id,
                        this.blockNode.current,
                      )
                  : null
              }
              className={cx(`block ${type}`, { selected: this.props.selected })}
              style={{ outline: 'none' }}
              ref={this.blockNode}
              // The tabIndex is required for the keyboard navigation
              /* eslint-disable jsx-a11y/no-noninteractive-tabindex */
              tabIndex={!blockHasOwnFocusManagement ? -1 : null}
            >
              <Block {...this.props} blockNode={this.blockNode} />
            </div>
          ) : (
            <div
              role="presentation"
              onKeyDown={e =>
                this.props.handleKeyDown(
                  e,
                  this.props.index,
                  this.props.id,
                  this.blockNode.current,
                )
              }
              className={cx(`block ${type}`, { selected: this.props.selected })}
              style={{ outline: 'none' }}
              ref={this.blockNode}
              // The tabIndex is required for the keyboard navigation
              tabIndex={-1}
            >
              {this.props.intl.formatMessage(messages.unknownBlock, {
                block: type,
              })}
            </div>
          )}
          {selected && !includes(blocks.requiredBlocks, type) && (
            <Button
              icon
              basic
              onClick={() => this.props.onDeleteBlock(id)}
              className="delete-button"
              aria-label="delete"
            >
              <Icon name={trashSVG} size="18px" />
            </Button>
          )}
        </div>,
      ),
    );
  }
}

export default compose(
  injectIntl,
  withObjectBrowser,
  DropTarget(ItemTypes.ITEM, itemTarget, connect => ({
    connectDropTarget: connect.dropTarget(),
  })),
  DragSource(ItemTypes.ITEM, itemSource, (connect, monitor) => ({
    connectDragSource: connect.dragSource(),
    connectDragPreview: connect.dragPreview(),
  })),
)(Edit);
