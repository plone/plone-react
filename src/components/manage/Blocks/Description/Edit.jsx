/**
 * Edit description tile.
 * @module components/manage/Tiles/Description/Edit
 */

import React, { Component } from 'react';
import { Map } from 'immutable';
import PropTypes from 'prop-types';
import { stateFromHTML } from 'draft-js-import-html';
import { Editor, DefaultDraftBlockRenderMap, EditorState } from 'draft-js';
import { defineMessages, injectIntl } from 'react-intl';
import cx from 'classnames';

const messages = defineMessages({
  description: {
    id: 'Add a description…',
    defaultMessage: 'Add a description…',
  },
});

const blockRenderMap = Map({
  unstyled: {
    element: 'div',
  },
});

const extendedBlockRenderMap = DefaultDraftBlockRenderMap.merge(blockRenderMap);

/**
 * Edit description tile class.
 * @class Edit
 * @extends Component
 */
class Edit extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    properties: PropTypes.objectOf(PropTypes.any).isRequired,
    selected: PropTypes.bool.isRequired,
    tile: PropTypes.string.isRequired,
    index: PropTypes.number.isRequired,
    onChangeField: PropTypes.func.isRequired,
    onSelectTile: PropTypes.func.isRequired,
    onDeleteTile: PropTypes.func.isRequired,
    onAddTile: PropTypes.func.isRequired,
    onFocusPreviousTile: PropTypes.func.isRequired,
    onFocusNextTile: PropTypes.func.isRequired,
  };

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Component properties
   * @constructs WysiwygEditor
   */
  constructor(props) {
    super(props);

    if (!__SERVER__) {
      let editorState;
      if (props.properties && props.properties.description) {
        const contentState = stateFromHTML(props.properties.description);
        editorState = EditorState.createWithContent(contentState);
      } else {
        editorState = EditorState.createEmpty();
      }
      this.state = { editorState, focus: false };
    }

    this.onChange = this.onChange.bind(this);
  }

  /**
   * Component did mount lifecycle method
   * @method componentDidMount
   * @returns {undefined}
   */
  componentDidMount() {
    if (this.node) {
      this.node._onBlur = () => this.setState({ focus: false });
      this.node._onFocus = () => this.setState({ focus: true });
    }
  }

  /**
   * Component will receive props
   * @method componentWillReceiveProps
   * @param {Object} nextProps Next properties
   * @returns {undefined}
   */
  componentWillReceiveProps(nextProps) {
    if (
      nextProps.properties.description &&
      this.props.properties.description !== nextProps.properties.description &&
      !this.state.focus
    ) {
      const contentState = stateFromHTML(nextProps.properties.description);
      this.setState({
        editorState: nextProps.properties.description
          ? EditorState.createWithContent(contentState)
          : EditorState.createEmpty(),
      });
    }

    if (!this.props.selected && nextProps.selected) {
      this.node.focus();
      this.setState({ focus: true });
    }
  }

  /**
   * Change handler
   * @method onChange
   * @param {object} editorState Editor state.
   * @returns {undefined}
   */
  onChange(editorState) {
    this.setState({ editorState }, () => {
      this.props.onChangeField(
        'description',
        editorState.getCurrentContent().getPlainText(),
      );
    });
  }

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    if (__SERVER__) {
      return <div />;
    }
    return (
      <div
        className={cx('tile description', { selected: this.props.selected })}
      >
        <Editor
          onChange={this.onChange}
          editorState={this.state.editorState}
          blockRenderMap={extendedBlockRenderMap}
          handleReturn={() => {
            this.props.onSelectTile(
              this.props.onAddTile('text', this.props.index + 1),
            );
            return 'handled';
          }}
          handleKeyCommand={(command, editorState) => {
            if (
              command === 'backspace' &&
              editorState.getCurrentContent().getPlainText().length === 0
            ) {
              this.props.onDeleteTile(this.props.tile, true);
            }
          }}
          placeholder={this.props.intl.formatMessage(messages.description)}
          blockStyleFn={() => 'documentDescription'}
          onUpArrow={() => {
            const selectionState = this.state.editorState.getSelection();
            const { editorState } = this.state;
            if (
              editorState
                .getCurrentContent()
                .getBlockMap()
                .first()
                .getKey() === selectionState.getFocusKey()
            ) {
              this.props.onFocusPreviousTile(this.props.tile, this.node);
            }
          }}
          onDownArrow={() => {
            const selectionState = this.state.editorState.getSelection();
            const { editorState } = this.state;
            if (
              editorState
                .getCurrentContent()
                .getBlockMap()
                .last()
                .getKey() === selectionState.getFocusKey()
            ) {
              this.props.onFocusNextTile(this.props.tile, this.node);
            }
          }}
          ref={node => {
            this.node = node;
          }}
        />
      </div>
    );
  }
}

export default injectIntl(Edit);
