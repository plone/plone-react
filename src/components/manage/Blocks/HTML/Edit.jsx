/**
 * Edit html block.
 * @module components/manage/Blocks/HTML/Edit
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Editor from 'react-simple-code-editor';
import { highlight, languages } from 'prismjs/components/prism-core';
import 'prismjs/components/prism-markup';
import { Button } from 'semantic-ui-react';
import pretty from 'pretty';

import { Icon } from '../../../../components';
import showSVG from '../../../../icons/show.svg';
import clearSVG from '../../../../icons/clear.svg';
import codeSVG from '../../../../icons/code.svg';

/**
 * Edit html block class.
 * @class Edit
 * @extends Component
 */
class Edit extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    selected: PropTypes.bool.isRequired,
    block: PropTypes.string.isRequired,
    index: PropTypes.number.isRequired,
    data: PropTypes.objectOf(PropTypes.any).isRequired,
    onChangeBlock: PropTypes.func.isRequired,
    onSelectBlock: PropTypes.func.isRequired,
    onDeleteBlock: PropTypes.func.isRequired,
    handleKeyDown: PropTypes.func.isRequired,
  };

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Component properties
   */
  constructor(props) {
    super(props);
    this.state = {
      code: this.props.data.html || '',
      isPreview: false,
    };
    this.onChangeCode = this.onChangeCode.bind(this);
    this.onPreview = this.onPreview.bind(this);
    this.onCodeEditor = this.onCodeEditor.bind(this);
  }

  /**
   * Component will receive props
   * @method componentDidMount
   * @returns {undefined}
   */
  componentDidMount() {
    if (this.props.selected) {
      this.codeEditor._input.focus();
    }
  }

  /**
   * Component will receive props
   * @method componentWillReceiveProps
   * @param {Object} nextProps Next properties
   * @returns {undefined}
   */
  UNSAFE_componentWillReceiveProps(nextProps) {
    if (nextProps.selected) {
      this.codeEditor._input.focus();
    }
  }

  /**
   * Change html handler
   * @method onChangeCode
   * @param {string} code New value html
   * @returns {undefined}
   */
  onChangeCode(code) {
    this.props.onChangeBlock(this.props.block, {
      ...this.props.data,
      html: code,
    });
    this.setState({ code });
  }

  /**
   * Preview mode handler
   * @method onPreview
   * @returns {undefined}
   */
  onPreview() {
    this.setState({
      isPreview: !this.state.isPreview,
      code: pretty(this.state.code),
    });
  }

  /**
   * Code Editor mode handler
   * @method onPreview
   * @returns {undefined}
   */
  onCodeEditor() {
    this.setState({ isPreview: !this.state.isPreview });
  }

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    return (
      <>
        {this.props.selected && !!this.state.code && (
          <div className="toolbar">
            <Button.Group>
              <Button
                icon
                basic
                aria-label="Source"
                active={!this.state.isPreview}
                onClick={this.onCodeEditor}
              >
                <Icon name={codeSVG} size="24px" />
              </Button>
            </Button.Group>
            <Button.Group>
              <Button
                icon
                basic
                aria-label="Preview"
                active={this.state.isPreview}
                onClick={this.onPreview}
              >
                <Icon name={showSVG} size="24px" />
              </Button>
            </Button.Group>
            <div className="separator" />
            <Button.Group>
              <Button icon basic onClick={() => this.onChangeCode('')}>
                <Icon name={clearSVG} size="24px" color="#e40166" />
              </Button>
            </Button.Group>
          </div>
        )}
        {this.state.isPreview && (
          <div dangerouslySetInnerHTML={{ __html: this.state.code }} />
        )}
        {!this.state.isPreview && (
          <Editor
            value={this.state.code}
            placeholder={`<p>Add some HTML here</p>`}
            onValueChange={code => this.onChangeCode(code)}
            highlight={code => highlight(code, languages.html)}
            padding={8}
            className="html-editor"
            ref={node => {
              this.codeEditor = node;
            }}
          />
        )}
      </>
    );
  }
}

export default Edit;
