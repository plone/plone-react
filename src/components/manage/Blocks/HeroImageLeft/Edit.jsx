/**
 * Edit Hero block.
 * @module components/manage/Blocks/Image/Edit
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { Map } from 'immutable';
import { readAsDataURL } from 'promise-file-reader';
import { Button, Dimmer, Loader, Message } from 'semantic-ui-react';
import { defineMessages, injectIntl } from 'react-intl';
import cx from 'classnames';

import { flattenToAppURL, getBaseUrl } from '@plone/volto/helpers';
import { createContent } from '@plone/volto/actions';
import { Icon } from '@plone/volto/components';

import clearSVG from '@plone/volto/icons/clear.svg';

import loadable from '@loadable/component';
const LibDraftJs = loadable.lib(() => import('draft-js'));
const LibDraftJsImportHtml = loadable.lib(() => import('draft-js-import-html'));

const messages = defineMessages({
  title: {
    id: 'Title',
    defaultMessage: 'Title',
  },
  description: {
    id: 'Description',
    defaultMessage: 'Description',
  },
  placeholder: {
    id: 'Upload a new image',
    defaultMessage: 'Upload a new image',
  },
});

const blockTitleRenderMap = Map({
  unstyled: {
    element: 'h1',
  },
});

const blockDescriptionRenderMap = Map({
  unstyled: {
    element: 'div',
  },
});

/**
 * Edit image block class.
 * @class Edit
 * @extends Component
 */
class Edit extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    selected: PropTypes.bool.isRequired,
    block: PropTypes.string.isRequired,
    index: PropTypes.number.isRequired,
    data: PropTypes.objectOf(PropTypes.any).isRequired,
    content: PropTypes.objectOf(PropTypes.any).isRequired,
    request: PropTypes.shape({
      loading: PropTypes.bool,
      loaded: PropTypes.bool,
    }).isRequired,
    pathname: PropTypes.string.isRequired,
    onChangeBlock: PropTypes.func.isRequired,
    onSelectBlock: PropTypes.func.isRequired,
    onDeleteBlock: PropTypes.func.isRequired,
    onFocusPreviousBlock: PropTypes.func.isRequired,
    onFocusNextBlock: PropTypes.func.isRequired,
    handleKeyDown: PropTypes.func.isRequired,
    createContent: PropTypes.func.isRequired,
  };

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Component properties
   * @constructs WysiwygEditor
   */
  constructor(props) {
    super(props);

    this.onUploadImage = this.onUploadImage.bind(this);
    this.state = {
      uploading: false,
    };

    if (!__SERVER__) {
      this.state = {
        uploading: false,
        titleEditorState: null,
        descriptionEditorState: null,
        currentFocused: 'title',
      };
    }

    this.onChangeTitle = this.onChangeTitle.bind(this);
    this.onChangeDescription = this.onChangeDescription.bind(this);
  }

  extendedBlockRenderMap = null;
  extendedDescripBlockRenderMap = null;

  /**
   * Component will receive props
   * @method componentWillReceiveProps
   * @param {Object} nextProps Next properties
   * @returns {undefined}
   */
  UNSAFE_componentWillReceiveProps(nextProps) {
    if (
      this.props.request.loading &&
      nextProps.request.loaded &&
      this.state.uploading
    ) {
      this.setState({
        uploading: false,
      });
      this.props.onChangeBlock(this.props.block, {
        ...this.props.data,
        url: nextProps.content['@id'],
      });
    }

    if (nextProps.selected !== this.props.selected) {
      if (this.state.currentFocused === 'title') {
        this.titleEditor.focus();
      } else {
        this.descriptionEditor.focus();
      }
    }
  }

  /**
   * Change Title handler
   * @method onChangeTitle
   * @param {object} titleEditorState Editor state.
   * @returns {undefined}
   */
  onChangeTitle(titleEditorState) {
    this.setState({ titleEditorState }, () => {
      this.props.onChangeBlock(this.props.block, {
        ...this.props.data,
        title: titleEditorState.getCurrentContent().getPlainText(),
      });
    });
  }

  /**
   * Change Description handler
   * @method onChangeDescription
   * @param {object} descriptionEditorState Editor state.
   * @returns {undefined}
   */
  onChangeDescription(descriptionEditorState) {
    this.setState({ descriptionEditorState }, () => {
      this.props.onChangeBlock(this.props.block, {
        ...this.props.data,
        description: descriptionEditorState.getCurrentContent().getPlainText(),
      });
    });
  }

  /**
   * Upload image handler
   * @method onUploadImage
   * @returns {undefined}
   */
  onUploadImage({ target }) {
    const file = target.files[0];
    this.setState({
      uploading: true,
    });
    readAsDataURL(file).then((data) => {
      const fields = data.match(/^data:(.*);(.*),(.*)$/);
      this.props.createContent(getBaseUrl(this.props.pathname), {
        '@type': 'Image',
        image: {
          data: fields[3],
          encoding: fields[2],
          'content-type': fields[1],
          filename: file.name,
        },
      });
    });
  }

  firstRender = true;

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    if (__SERVER__) {
      return <div />;
    }
    const placeholder =
      this.props.data.placeholder ||
      this.props.intl.formatMessage(messages.placeholder);
    return (
      <LibDraftJs>
        {({ Editor, EditorState, DefaultDraftBlockRenderMap }) => {
          return (
            <LibDraftJsImportHtml>
              {({ stateFromHTML }) => {
                if (!__SERVER__ && this.firstRender) {
                  this.firstRender = false;

                  this.extendedBlockRenderMap = DefaultDraftBlockRenderMap.merge(
                    blockTitleRenderMap,
                  );

                  this.extendedDescripBlockRenderMap = DefaultDraftBlockRenderMap.merge(
                    blockDescriptionRenderMap,
                  );

                  let titleEditorState;
                  let descriptionEditorState;
                  if (this.props.data && this.props.data.title) {
                    titleEditorState = EditorState.createWithContent(
                      stateFromHTML(this.props.data.title),
                    );
                  } else {
                    titleEditorState = EditorState.createEmpty();
                  }
                  if (this.props.data && this.props.data.description) {
                    descriptionEditorState = EditorState.createWithContent(
                      stateFromHTML(this.props.data.description),
                    );
                  } else {
                    descriptionEditorState = EditorState.createEmpty();
                  }

                  this.setState({
                    uploading: false,
                    titleEditorState,
                    descriptionEditorState,
                    currentFocused: 'title',
                  });
                }

                return (
                  this.state.titleEditorState &&
                  this.state.descriptionEditorState && (
                    <div
                      className={cx('block hero', {
                        selected: this.props.selected,
                      })}
                    >
                      {this.props.selected && !!this.props.data.url && (
                        <div className="toolbar">
                          <Button.Group>
                            <Button
                              icon
                              basic
                              onClick={() =>
                                this.props.onChangeBlock(this.props.block, {
                                  ...this.props.data,
                                  url: '',
                                })
                              }
                            >
                              <Icon
                                name={clearSVG}
                                size="24px"
                                color="#e40166"
                              />
                            </Button>
                          </Button.Group>
                        </div>
                      )}
                      <div className="block-inner-wrapper">
                        {this.props.data.url ? (
                          <img
                            className="hero-image"
                            src={`${flattenToAppURL(
                              this.props.data.url,
                            )}/@@images/image`}
                            alt=""
                          />
                        ) : (
                          <div className="image-add">
                            <Message className="image-message">
                              {this.state.uploading && (
                                <Dimmer active>
                                  <Loader indeterminate>Uploading image</Loader>
                                </Dimmer>
                              )}
                              <center>
                                <h4>Image</h4>
                                <p>{placeholder}</p>
                                <p>
                                  <label className="ui button file">
                                    Browse
                                    <input
                                      type="file"
                                      onChange={this.onUploadImage}
                                      style={{ display: 'none' }}
                                    />
                                  </label>
                                </p>
                              </center>
                            </Message>
                          </div>
                        )}
                        <div className="hero-body">
                          <Editor
                            ref={(node) => {
                              this.titleEditor = node;
                              if (this.props.selected && this.titleEditor) {
                                this.titleEditor.focus();
                              }
                            }}
                            onChange={this.onChangeTitle}
                            editorState={this.state.titleEditorState}
                            blockRenderMap={this.extendedBlockRenderMap}
                            handleReturn={() => true}
                            placeholder={this.props.intl.formatMessage(
                              messages.title,
                            )}
                            blockStyleFn={() => 'title-editor'}
                            onUpArrow={() => {
                              const selectionState = this.state.titleEditorState.getSelection();
                              const { titleEditorState } = this.state;
                              if (
                                titleEditorState
                                  .getCurrentContent()
                                  .getBlockMap()
                                  .first()
                                  .getKey() === selectionState.getFocusKey()
                              ) {
                                this.props.onFocusPreviousBlock(
                                  this.props.block,
                                  this.props.blockNode.current,
                                );
                              }
                            }}
                            onDownArrow={() => {
                              const selectionState = this.state.titleEditorState.getSelection();
                              const { titleEditorState } = this.state;
                              // if (
                              //   titleEditorState
                              //     .getCurrentContent()
                              //     .getBlockMap()
                              //     .last()
                              //     .getKey() === selectionState.getFocusKey()
                              // ) {
                              this.setState(
                                () => ({
                                  currentFocused: 'description',
                                }),
                                () => {
                                  debugger;
                                  this.descriptionEditor.focus();
                                },
                              );
                              //   }
                              // }}
                            }}
                          />
                          <Editor
                            ref={(node) => {
                              this.descriptionEditor = node;
                            }}
                            onChange={this.onChangeDescription}
                            editorState={this.state.descriptionEditorState}
                            blockRenderMap={this.extendedDescripBlockRenderMap}
                            handleReturn={() => true}
                            placeholder={this.props.intl.formatMessage(
                              messages.description,
                            )}
                            blockStyleFn={() => 'description-editor'}
                            onUpArrow={() => {
                              const selectionState = this.state.descriptionEditorState.getSelection();
                              const currentCursorPosition = selectionState.getStartOffset();

                              if (currentCursorPosition === 0) {
                                this.setState(
                                  () => ({
                                    currentFocused: 'title',
                                  }),
                                  () => {
                                    this.titleEditor.focus();
                                  },
                                );
                              }
                            }}
                            onDownArrow={() => {
                              const selectionState = this.state.descriptionEditorState.getSelection();
                              const { descriptionEditorState } = this.state;
                              const currentCursorPosition = selectionState.getStartOffset();
                              const blockLength = descriptionEditorState
                                .getCurrentContent()
                                .getFirstBlock()
                                .getLength();

                              if (currentCursorPosition === blockLength) {
                                this.props.onFocusNextBlock(
                                  this.props.block,
                                  this.props.blockNode.current,
                                );
                              }
                            }}
                          />
                        </div>
                      </div>
                    </div>
                  )
                );
              }}
            </LibDraftJsImportHtml>
          );
        }}
      </LibDraftJs>
    );
  }
}

export default compose(
  injectIntl,
  connect(
    (state) => ({
      request: state.content.create,
      content: state.content.data,
    }),
    { createContent },
  ),
)(Edit);
