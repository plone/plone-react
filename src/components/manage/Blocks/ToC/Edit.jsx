/**
 * Edit toc block.
 * @module components/manage/Blocks/ToC/Edit
 */

import React from 'react';
import PropTypes from 'prop-types';

import View from '@plone/volto/components/manage/Blocks/ToC/View';

/**
 * Edit toc block class.
 * @class Edit
 * @extends Component
 */
const Edit = ({ properties }) => <View properties={properties} />;

/**
 * Property types.
 * @property {Object} propTypes Property types.
 * @static
 */
Edit.propTypes = {
  properties: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default Edit;
