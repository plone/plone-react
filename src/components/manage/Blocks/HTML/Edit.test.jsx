import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';

import Edit from './Edit';

const mockStore = configureStore();

test('renders an edit html block component', () => {
  const store = mockStore({
    content: {
      create: {},
      data: {},
    },
    intl: {
      locale: 'en',
      messages: {},
    },
  });
  const component = renderer.create(
    <Provider store={store}>
      <Edit
        data={{ html: '<h1></h1>' }}
        selected={false}
        block="1234"
        onChangeBlock={() => {}}
        onSelectBlock={() => {}}
        onDeleteBlock={() => {}}
        onFocusPreviousBlock={() => {}}
        onFocusNextBlock={() => {}}
        handleKeyDown={() => {}}
        index={1}
      />
    </Provider>,
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
