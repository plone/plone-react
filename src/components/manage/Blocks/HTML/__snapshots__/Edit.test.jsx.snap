// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an edit html block component 1`] = `
<div
  class="html-editor"
  style="position: relative; text-align: left; white-space: pre-wrap; word-break: keep-all; box-sizing: border-box; padding: 0px;"
>
  <textarea
    autocapitalize="off"
    autocomplete="off"
    autocorrect="off"
    class="npm__react-simple-code-editor__textarea "
    data-gramm="false"
    placeholder="<p>Add some HTML here</p>"
    spellcheck="false"
    style="box-sizing: inherit; display: inherit; font-family: inherit; font-style: inherit; font-variant-ligatures: inherit; font-weight: inherit; letter-spacing: inherit; line-height: inherit; tab-size: inherit; text-indent: inherit; text-rendering: inherit; text-transform: inherit; white-space: inherit; word-break: inherit; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; margin: 0px; border: 0px; resize: none; background: none; overflow: hidden; padding: 8px 8px 8px 8px;"
  >
    &lt;h1&gt;&lt;/h1&gt;
  </textarea>
  <pre
    aria-hidden="true"
    style={
      Object {
        "border": 0,
        "boxSizing": "inherit",
        "display": "inherit",
        "fontFamily": "inherit",
        "fontSize": "inherit",
        "fontStyle": "inherit",
        "fontVariantLigatures": "inherit",
        "fontWeight": "inherit",
        "letterSpacing": "inherit",
        "lineHeight": "inherit",
        "margin": 0,
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
        "pointerEvents": "none",
        "position": "relative",
        "tabSize": "inherit",
        "textIndent": "inherit",
        "textRendering": "inherit",
        "textTransform": "inherit",
        "whiteSpace": "inherit",
        "wordBreak": "inherit",
      }
    }
  />
  <style
    type="text/css"
  >

/**
 * Reset the text fill color so that placeholder is visible
 */
.npm__react-simple-code-editor__textarea:empty {
  -webkit-text-fill-color: inherit !important;
}

/**
 * IE doesn't support '-webkit-text-fill-color'
 * So we use 'color: transparent' to make the text transparent on IE
 * Unlike other browsers, it doesn't affect caret color in IE
 */
.npm__react-simple-code-editor__textarea-ie {
  color: transparent !important;
}

.npm__react-simple-code-editor__textarea-ie::selection {
  background-color: #accef7 !important;
  color: transparent !important;
}

  </style>
</div>
`;
